File: src/main/java/functor/Functor.java
package functor;

import common.DataClass;

import java.util.function.Function;

public interface Functor<A> {
    <B> Functor<B> fmap(Function<A, B> f);
}

abstract class Maybe<A> extends DataClass implements Functor<A> {
}

class Just<A> extends Maybe<A> {

    private final A value;

    Just(A value) {
        this.value = value;
    }

    @Override
    public <B> Functor<B> fmap(Function<A, B> f) {
        return new Just(f.apply(value));
    }
}

class Nothing<A> extends Maybe<A> {

    @Override
    public <B> Functor<B> fmap(Function<A, B> f) {
        return new Nothing<>();
    }
}

File: src/main/kotlin/root_k/functor/functor.kt
package root_k.functor

import root_k.functor.Maybe.Just
import root_k.functor.Maybe.None

sealed class Maybe<out A> {
    object None : Maybe<Nothing>()
    data class Just<out A>(val value: A) : Maybe<A>()

    infix fun <B> fmap(f: (A) -> B): Maybe<B> = when (this) {
        is None -> this
        is Just -> Just(f(value))
    }
}

fun main(args: Array<String>) {
    None fmap { print("this will not be applied") } // None

    Just(5) fmap { it + 1 } fmap { it + 2 } // Just 8
}
